src/psyclone/psyGen.py

```diff
@@ -1768,15 +1803,22 @@ def gen_code(self, parent):
         for entity in self._children:
             entity.gen_code(parent)
 
-        if self.opencl:
-            # Ensure we block at the end of the invoke to ensure all
-            # kernels have completed before we return.
-            # This code ASSUMES only the first command queue is used for
-            # executing kernels.
+        if self.opencl and self._opencl_options['end_barrier']:
+
             parent.add(CommentGen(parent,
                                   " Block until all kernels have finished"))
-            parent.add(AssignGen(parent, lhs=flag,
-                                 rhs="clFinish(" + qlist + "(1))"))
+
+            # We need a clFinish for all the queues in the implementation
+            opencl_num_queues = 1
+            for kern in self.coded_kernels():
+                opencl_num_queues = max(
+                    opencl_num_queues,
+                    kern._opencl_options['queue_number'])
```

Comment:
pylint complains about this reference to `_opencl_options` - I guess we need a getter.

Reply:
Added a property to access opencl_options